tests/test_functools.py

```diff
@@ -0,0 +1,294 @@
+            await checkpoint()
+            counter += 1
+            return counter
+
+        # These two calls should be cached with different keys
+        assert await func(1, "y", 2) == 1
+        assert await func(1, y=2) == 2
+
+    async def test_cache_same_function_twice(self) -> None:
+        counter = 0
+
+        async def func() -> int:
+            nonlocal counter
+            await checkpoint()
+            counter += 1
+            return counter
+
+        cached_1 = lru_cache()(func)
+        cached_2 = lru_cache()(func)
+
+        # This should yield two cache misses
+        assert await cached_1() == 1
+        assert await cached_2() == 2
+
+
+class TestReduce:
+    async def test_not_iterable(self) -> None:
+        with pytest.raises(
+            TypeError, match="argument 2 must be an iterable or async iterable"
+        ):
+            await reduce(lambda x, y: x + y, 1)  # type: ignore[call-overload]
+
+    async def test_no_initial(self) -> None:
+        async def func(x: int, y: int) -> int:
+            await checkpoint()
+            return x + y
+
+        assert await reduce(func, [1, 2, 3]) == 6
+
+    async def test_has_initial(self) -> None:
+        async def func(x: int, y: str) -> int:
+            await checkpoint()
+            return x + int(y)
+
+        assert await reduce(func, ["1", "2", "3"], 2) == 8
+
+    async def test_empty_iter_no_initial(self) -> None:
+        with pytest.raises(
+            TypeError, match=r"reduce\(\) of empty sequence with no initial value"
+        ):
+            await reduce(lambda x, y: x + y, [])
+
+    async def test_asynciter_no_initial(self) -> None:
+        async def func(x: int, y: int) -> int:
+            await checkpoint()
+            return x + y
+
+        async def asyncgen() -> AsyncIterator[int]:
+            yield 1
+            yield 2
+            yield 3
+
+        assert await reduce(func, asyncgen()) == 6
+
+    async def test_asynciter_has_initial(self) -> None:
+        async def func(x: int, y: str) -> int:
+            await checkpoint()
+            return x + int(y)
+
+        async def asyncgen() -> AsyncIterator[str]:
+            yield "1"
+            yield "2"
+            yield "3"
+
+        assert await reduce(func, asyncgen(), 2) == 8
+
+    async def test_empty_iterable_no_initial(self) -> None:
+        with pytest.raises(
+            TypeError, match=r"reduce\(\) of empty sequence with no initial value"
+        ):
+            await reduce(lambda x, y: x + y, ())
+
+    async def test_empty_async_iterable_no_initial(self) -> None:
+        class AIter:
+            def __aiter__(self) -> AIter:
+                return self
+
+            async def __anext__(self) -> NoReturn:
+                raise StopAsyncIteration
+
+        with pytest.raises(
+            TypeError, match=r"reduce\(\) of empty sequence with no initial value"
+        ):
+            await reduce(lambda x, y: x + y, AIter())
```

Comment:
A bit of a pathological case, but `reduce()` doesn't currently visit a checkpoint when the iterable is empty. I think that we should add a checkpoint if the iterable is empty.

```suggestion
            await reduce(lambda x, y: x + y, AIter())

    async def test_checkpoints_empty_iterable(self) -> None:
        async def func(x: int, y: int) -> int:
            await checkpoint()
            return x + y

        with CancelScope() as cs:
            cs.cancel()
            with pytest.raises(get_cancelled_exc_class()):
                await reduce(func, [], 1)

    async def test_checkpoints_empty_async_iterable(self) -> None:
        async def func(x: int, y: int) -> int:
            await checkpoint()
            return x + y

        class AIter:
            def __aiter__(self) -> AIter:
                return self

            async def __anext__(self) -> NoReturn:
                raise StopAsyncIteration

        with CancelScope() as cs:
            cs.cancel()
            with pytest.raises(get_cancelled_exc_class()):
                await reduce(func, AIter(), 1)
```

Reply:
Yeah. If there are no checkpoints in the user's function then there is a bug in the user's function though.

In contrast: if `await reduce(function_that_checkpoints, ...)` returns without visiting a checkpoint then I think there is a bug in `reduce`. I don't think that users should have to write

```python
async def foo():
    if not iter_is_empty:
        return await reduce(function_that_checkpoints, iter_, init)
    else:
        # We can't just `return await reduce(function_that_checkpoints, iter_, init)`
        # because `reduce` doesn't checkpoint in this case. We need to add a checkpoint.
        await anyio.lowlevel.checkpoint()
        return init
```

instead of just `return await reduce(function_that_checkpoints, iter_, init)`.

Reply:
All that said, I've added code to ensure that there is a checkpoint even in the case of an initial value and an empty iterable.

Reply:
Thanks, I greatly appreciate you taking the time to write that longer message. I see how this problem is difficult. It reminds me of the saying "perfect is the enemy of good." Thank you for pushing for SC/level-cancellation/other ease-of-use improvements in asyncio-land.